string dictionary
---
string final_result
---
string partial_result